Add HTMX and Alpine.js alongside helpful utilities, common js_entry point

biyeun opened the pull request:
## Product Description
The time has come! We are officially including [HTMX](https://htmx.org/) and [Alpine](https://alpinejs.dev/) in our JavaScript dependencies.

I didn't want to just drop these dependencies into HQ without some examples and documentation providing structure for how we want to use these libraries in our codebase. With such a large team, structure is immensely important. Hopefully, these examples and corresponding docs in the Styleguide will provide a good starting point for us!

### Example 1: CrispyForms usage
This example uses Alpine to sprinkle interactivity into a Crispy Form—a very common use case for HQ. It also uses
HTMX to asynchronously load that form to a "host" view and process the response from that form.

![Screenshot 2024-10-16 at 2 24 03 PM](https://github.com/user-attachments/assets/d09d9f21-3990-4c82-8bc7-b0f84127ec4a)
[view on staging](https://staging.commcarehq.org/styleguide/demo/htmx_alpine_form/) (must be logged in)

### Example 2: To-Do List
Taking inspiration from @snopoke's presentation at the GTD Summit, we have a simple To-Do list demoing the usage of `HtmxActionMixin` (explained in the technical summary). Additionally, this demos the usage of Alpine for a simple inline editing widget.

[view on staging](https://staging.commcarehq.org/styleguide/demo/htmx_todo/) (must be logged in)

![Screenshot 2024-10-16 at 2 23 44 PM](https://github.com/user-attachments/assets/d17f19a0-2003-497c-8500-7a9e230ce96e)
![Screenshot 2024-10-16 at 2 31 32 PM](https://github.com/user-attachments/assets/62d95f67-e29e-4e2c-a3e8-802be5ce43ca)
![Screenshot 2024-10-16 at 2 23 53 PM](https://github.com/user-attachments/assets/7e8b226d-28a2-4846-acd1-8bd005e15661)
(a demo of the HTMX request error modal can be triggered here as well!)

### Styleguide documentation to visually link everything together for the demos!
![Screenshot 2024-10-17 at 1 02 10 AM](https://github.com/user-attachments/assets/2575169b-5e2b-4978-b553-82f5f7755e25)

## Technical Summary
- Adds `htmx` and `alpine.js` as javascript dependencies
- adds a common `js_entry` point `hqwebapp/js/htmx_and_alpine` that can be used on a page for folks to quickly get started with HTMX and Alpine without writing any additional javascript files
- adds `HqHtmxActionMixin`, which can be mixed into django template views so that HTMX requests to that view can process and return partial template responses via methods decorated with `@hq_hx_action()`
- adds an example To-Do list illustrating the usage of `HqHtmxActionMixin`
- adds an example usage of alpine inside a crispy form for adding interactivity without extra javascript
- provides an example with an alternative to `HqHtmxActionMixin` for form interactions with HTMX that `FormView` that processes HTMX requests and is hosted on a regular template view. Note, it's possible to also have that template view use `HqHtmxActionMixin` for additional non-form HTMX activity for more complex usage
- adds a utility which inserts the CSRF token into HTMX requests
- adds a utility to retry HTMX requests
- adds a utility to handle HTMX errors and generate a modal (with alpine) to relay those errors to the user
- adds guidance for accessing Alpine globals (e.g. `Alpine.data()`) for more complex pages
- documentation in styleguide going over the two examples above and an quick start suggestion
- `HqHtmxDebugMixin` added for simulating server issues during development of views using `HqHtmxActionMixin`

## Safety Assurance

### Safety story
Only introducing new dependencies and examples / documentation. Not affecting any active pages.

### Automated test coverage
N/A

### QA Plan
N/A


### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

biyeun committed:
add HTMX to js dependencies

biyeun committed:
add Alpine.js to javascript dependencies

biyeun committed:
add HtmxActionMixin and support for hq-hx-action attribute

biyeun committed:
configure HTMX requests to include the CSRF token

biyeun committed:
add a utility for triggering HTMX error modal

biyeun committed:
add utility for retrying HTMX requests

biyeun committed:
stitch everything together to create a common entry point for htmx + alpine to get going quickly

biyeun committed:
add CacheStore helper model for HTMX demos to prototype app

biyeun committed:
add example for using HTMX and Alpine with a simple Crispy Form

biyeun committed:
adds an example using HtmxActionMixin and alpine for inline editing

biyeun committed:
update the timeout to something more reasonable

nospame reviewed (approved):
Comments from today's product team code review. Makes sense on the whole, noting I (and other devs probably) will need to read through htmx and alpine docs for some particulars.

biyeun committed:
feedback: make retryPathCounts a const

biyeun committed:
improvement: reorder file so that HtmxActionMixin comes first

biyeun committed:
feedback: use constant instead of string with method='auto'

biyeun committed:
feedback: update naming of HtmxActionMixin and related

biyeun committed:
feedback: move Htmx debug tools into its own mixin

biyeun committed:
update response on errors to provide a reason so that HTMX displays the full message

biyeun committed:
feedback: return forbidden response if action isn't callable

biyeun committed:
feedback: not every HTMX evt is the same
make sure the correct details are passed to retryHtmxRequest

biyeun committed:
feedback: only retry on GET timeouts

biyeun committed:
feedback: clarify error codes, messages, and timeout units

biyeun committed:
feedback: default_value is renamed to initial_value

biyeun committed:
improvement: add help text to hint what the operative action is

biyeun committed:
rename templates and make paths shorter for htmx examples so that it's easier to follow and read

biyeun committed:
update comments

biyeun committed:
add django ace language support to styleguide

biyeun committed:
also disable checkbox when isSubmitting is true

biyeun committed:
add styleguide section for htmx and alpine

biyeun commented on corehq/apps/hqwebapp/static/hqwebapp/js/htmx_utils/retry_request.js:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    // Prepare the context for the htmx request
+    const context = {
```
updated the signature of the function to make things easier to read. but, yes, no easy retry option with HTMX sadly

biyeun committed:
feedback: clarify timeout units and value

biyeun committed:
Merge branch 'master' of github.com:dimagi/commcare-hq into bmb/htmx-and-alpine

biyeun commented:
Thank you @nospame and @mjriley for your group review today! I addressed the feedback and added a styleguide page which will be available [here](https://staging.commcarehq.org/styleguide/b5/htmx_alpine/) as soon as the next staging deploy completes. :)

snopoke reviewed (approved):
Amazing work @biyeun
